### ISSUE TITLE
Have Dusk Download Files?

### ISSUE THREAD

**Issue by InstanceOfMichael** [laravel/dusk#486]
I have a site where clicking a button will download a CSV file (generated clientside, it's not a link). However, I can't get laravel dusk to download the file. I found https://github.com/SeleniumHQ/selenium/issues/5159 but wasn't sure how to translate it into PHP or Laravel Dusk.

**Comment by InstanceOfMichael** [comment]
It looks like works if I don't run in `--headless` mode. How do you make downloads work in headless mode?

**Comment by staudenmeir** [comment]
This works with Chrome 65 and ChromeDriver 2.36:

    $this->browse(function (Browser $browser) {
        $url = 'http://localhost:9515';
        $uri = '/session/' . $browser->driver->getSessionID() . '/chromium/send_command';
        $body = [
            'cmd' => 'Page.setDownloadBehavior',
            'params' => ['behavior' => 'allow', 'downloadPath' => '/your/download/path']
        ];
        (new \GuzzleHttp\Client())->post($url . $uri, ['body' => json_encode($body)]);

        $browser->visit('/');
    });

The value of `$url` must be the one used to create the driver in `driver()`.

Replace `/your/download/path` with the actual path.

This example requires the [`guzzlehttp/guzzle`](https://packagist.org/packages/guzzlehttp/guzzle) package to execute the POST request.

There are other solutions without external dependencies:  
https://stackoverflow.com/questions/5647461/how-do-i-send-a-post-request-with-php

**Comment by staudenmeir** [comment]
@InstanceOfMichael Does it work for you?

**Comment by chilio** [comment]
In my config, all works fine, if I disable --headless. 

So far, I have investigated the issue, and the problem is in the direct X lib, which is trying to mitigate certain capabilities of your gfx card. 
You can check this by invoking dxdiag in windows....
It is sad that it is not working out of the box, but you need to be prepared, that some dusk functionalities may fail, depending on your graphics card, even in headless mode....
So in other words, if you have latest CPU and GPU engines it should work....

**Comment by staudenmeir** [comment]
@chilio I think you are in the wrong issue ;-)

**Comment by chilio** [comment]
@staudenmeir you are right :)

**Comment by ankitjain2109** [comment]
Hi @staudenmeir , the code you gave is working fine when site directly download the file. But when the link to download file uses another tab it doesn't work. Is there any other possibility or param to define and download the file.

**Comment by InstanceOfMichael** [comment]
This worked, thanks!
